java.lang.NoSuchFieldError when mix Scala and Java types

## reproduction steps

using Scala (2.13.3 and Scala 2.12.12, SBT 1.4.1)
Minimal reproducible example: https://github.com/REASY/scala-compiler-bug

## problem
There is a small app https://github.com/REASY/scala-compiler-bug/blob/main/src/main/scala/example/Example.scala#L7 which suppose to print 
```
example.scala.Document.getMember(foo)
```
but instead it fails on runtime with the error `java.lang.NoSuchFieldError: document`:
```
$ sbt ";clean;cleanFiles;runMain example.Example"
[info] welcome to sbt 1.4.1 (Oracle Corporation Java 1.8.0_261)
[info] loading global plugins from C:\Users\User\.sbt\1.0\plugins
[info] loading settings for project scala-compiler-bug-build from plugins.sbt ...
[info] loading project definition from C:\Users\User\AppData\Local\Temp\scala-compiler-bug\project
[info] loading settings for project root from build.sbt ...
[info] set current project to compiler-bug (in build file:/C:/Users/User/AppData/Local/Temp/scala-compiler-bug/)
[success] Total time: 0 s, completed Oct 20, 2020 7:19:17 PM
[success] Total time: 0 s, completed Oct 20, 2020 7:19:18 PM
[info] compiling 2 Scala sources and 2 Java sources to C:\Users\User\AppData\Local\Temp\scala-compiler-bug\target\scala-2.13\classes ...
[info] done compiling
[info] running example.Example
[error] (run-main-0) java.lang.NoSuchFieldError: document
[error] java.lang.NoSuchFieldError: document
[error]         at example.Example$.main(Example.scala:9)
[error]         at example.Example.main(Example.scala)
[error]         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error]         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error]         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error]         at java.lang.reflect.Method.invoke(Method.java:498)
[error] stack trace is suppressed; run 'last Compile / bgRunMain' for the full output
[error] Nonzero exit code: 1
[error] (Compile / runMain) Nonzero exit code: 1
[error] Total time: 2 s, completed Oct 20, 2020 7:19:20 PM
```
![image](https://user-images.githubusercontent.com/5107562/96585032-3a60fc00-1309-11eb-850a-6d9ee325e236.png)

There is Java type [Window](https://github.com/REASY/scala-compiler-bug/blob/main/src/main/java/example/java/Window.java#L8) which has a public field `document` with Scala type [Document](https://github.com/REASY/scala-compiler-bug/blob/main/src/main/scala/example/scala/Document.scala). `Window` class is instantiated and accessed in runnable class https://github.com/REASY/scala-compiler-bug/blob/main/src/main/scala/example/Example.scala
```scala
package example

import example.java.Window

// To run via SBT use in shell: `sbt "runMain example.Example"`
object Example {
  def main(args: Array[String]): Unit = {
    val wnd = new Window()
    val res = wnd.document.getMember("foo").toString
    println(res)
  }
}
```

## Investigation
If I look to the generated bytecode using `javap` I can see that `example.java.Window` is looking as expected, `document` field's type is correct:
```
scala> :javap -c example.java.Window
Compiled from "Window.java"
public class example.java.Window {
  public final example.scala.Document document;

  public example.java.Window();
    Code:
       0: aload_0
       1: invokespecial #1                  // Method java/lang/Object."<init>":()V
       4: aload_0
       5: new           #2                  // class example/scala/Document
       8: dup
       9: invokespecial #3                  // Method example/scala/Document."<init>":()V
      12: putfield      #4                  // Field document:Lexample/scala/Document;
      15: return
}
```
On the other hand, the accessed type of document in `example.Example$.main` method is wrong, it is `example/java/Document` instead of `example/scala/Document`:
```
scala> :javap -c example.Example$
Compiled from "Example.scala"
public final class example.Example$ {
  public static final example.Example$ MODULE$;

  public static {};
    Code:
       0: new           #2                  // class example/Example$
       3: dup
       4: invokespecial #12                 // Method "<init>":()V
       7: putstatic     #14                 // Field MODULE$:Lexample/Example$;
      10: return

  public void main(java.lang.String[]);
    Code:
       0: new           #19                 // class example/java/Window
       3: dup
       4: invokespecial #20                 // Method example/java/Window."<init>":()V
       7: astore_2
       8: aload_2
       9: getfield      #24                 // Field example/java/Window.document:Lexample/java/Document;
      12: ldc           #26                 // String foo
      14: invokevirtual #32                 // Method example/java/Document.getMember:(Ljava/lang/String;)Ljava/lang/Object;
      17: invokevirtual #36                 // Method java/lang/Object.toString:()Ljava/lang/String;
      20: astore_3
      21: getstatic     #41                 // Field scala/Predef$.MODULE$:Lscala/Predef$;
      24: aload_3
      25: invokevirtual #45                 // Method scala/Predef$.println:(Ljava/lang/Object;)V
      28: return
}
```
![image](https://user-images.githubusercontent.com/5107562/96586686-8745d200-130b-11eb-9ffc-6fc4e2d83013.png)

TIL sbt compiler bridge has finite backward compatibility.
```
[info] Non-compiled module 'compiler-bridge_2.12' for Scala 2.12.0. Compiling...
/tmp/sbt_1d91b260/scala/ZincCompat.scala:19: error: type PlainNioFile is not a member of package scala.reflect.io
  type PlainNioFile = scala.reflect.io.PlainNioFile
```

--- a/src/compiler/scala/tools/nsc/typechecker/Contexts.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Contexts.scala
@@ -712,7 +712,7 @@ trait Contexts { self: Analyzer =>
         c(TypeConstructorAllowed) = false
 
       registerContext(c.asInstanceOf[analyzer.Context])
-      debuglog("[context] ++ " + c.unit + " / " + (if (tree == null) "" else tree.summaryString))
+      debuglog(s"[context] ++ ${c.unit} / ${if (tree == null) "" else tree.summaryString}")
       c
     }
 
@@ -1499,26 +1499,41 @@ trait Contexts { self: Analyzer =>
        *     as the reference, have lowest precedence. Also "root" imports added implicitly.
        */
       def foreignDefined = defSym.exists && thisContext.isPackageOwnedInDifferentUnit(defSym)  // SI-2458
-      // can an import at this depth possibly shadow the definition found in scope if any?
-      def importCanShadowAtDepth(imp: ImportInfo) = imp.depth > symbolDepth || (
-        if (thisContext.unit.isJava) imp.depth == symbolDepth && imp.isExplicitImport(name)
-        else foreignDefined
-      )
 
-      while (!impSym.exists && importCursor.imp1Exists && importCanShadowAtDepth(importCursor.imp1)) {
-        val (sel, sym) = lookupImport(imp1, requireExplicit = false)
-        impSel = sel
-        impSym = sym
-        if (!impSym.exists)
-          importCursor.advanceImp1Imp2()
+      // Find the first candidate import
+      def advanceCursorToNextImport(): Unit = {
+        val defIsLevel4 = foreignDefined
+        // can the import at this depth compete with the definition?
+        // If not, we can stop inspecting outer scopes (including more imports).
+        // A competing import can either shadow the definition or render it ambiguous.
+        //
+        @inline def importCanShadowAtDepth(imp: ImportInfo) = {
+          @inline def importCompetesWithDefinition =
+            if (thisContext.unit.isJava) imp.depth == symbolDepth && defIsLevel4
+            else defIsLevel4
+          imp.depth > symbolDepth || importCompetesWithDefinition
+        }
+
+        while (!impSym.exists && importCursor.imp1Exists && importCanShadowAtDepth(importCursor.imp1)) {
+          val javaRule = thisContext.unit.isJava && defIsLevel4
+          val (sel, sym) = lookupImport(imp1, requireExplicit = javaRule)
+          impSel = sel
+          impSym = sym
+          if (!impSym.exists)
+            importCursor.advanceImp1Imp2()
+        }
       }
+      advanceCursorToNextImport()
 
       val preferDef: Boolean = defSym.exists && (!impSym.exists || {
         // 4) root imported symbols have same (lowest) precedence as package-owned symbols in different compilation units.
         if (imp1.depth < symbolDepth && imp1.isRootImport && foreignDefined)
           true
         // 4) imported symbols have higher precedence than package-owned symbols in different compilation units.
-        else if (imp1.depth >= symbolDepth && foreignDefined)
+        //    except that in Java, the import must be "explicit" (level 2)
+        else if (thisContext.unit.isJava && imp1.depth == symbolDepth && foreignDefined)
+          !importCursor.imp1Explicit
+        else if (!thisContext.unit.isJava && imp1.depth >= symbolDepth && foreignDefined)
           false
         // Defined symbols take precedence over erroneous imports.
         else if (impSym.isError || impSym.name == nme.CONSTRUCTOR)
@@ -1531,8 +1546,8 @@ trait Contexts { self: Analyzer =>
           return ambiguousDefnAndImport(defSym.alternatives.head.owner, imp1)
       })
 
-      // If the defSym is at 4, and there is a def at 1 in scope, then the reference is ambiguous.
-      if (foreignDefined && !defSym.isPackage) {
+      // If the defSym is at 4, and there is a def at 1 in scope due to packaging, then the reference is ambiguous.
+      if (foreignDefined && !defSym.isPackage && !thisContext.unit.isJava) {
         val defSym0 = defSym
         val pre0    = pre
         val cx0     = cx
@@ -1919,7 +1934,7 @@ trait Contexts { self: Analyzer =>
     def sameDepth: Boolean = imp1.depth == imp2.depth
 
     private def imp2Exists = imp2Ctx.importOrNull != null
-    private def imp1Explicit = imp1 isExplicitImport name
+    def imp1Explicit = imp1 isExplicitImport name
     private def imp2Explicit = imp2 isExplicitImport name
   }
 

--- a/test/files/pos/t7232e/Foo.java
+++ b/test/files/pos/t7232e/Foo.java
@@ -1,0 +1,10 @@
+package pack;
+
+import java.util.List;   // does shadow
+import java.awt.*;       // doesn't shadow, keep looking
+import java.util.function.Supplier;  // irrelevant, keep looking
+
+public class Foo {
+    // should be java.util.List.
+    public static List list() { throw new Error(); }
+}

--- a/test/files/pos/t7232e/List.java
+++ b/test/files/pos/t7232e/List.java
@@ -1,0 +1,5 @@
+package pack;
+
+public class List {
+    public void packList() {}
+}

--- a/test/files/pos/t7232e/Test.scala
+++ b/test/files/pos/t7232e/Test.scala
@@ -1,0 +1,7 @@
+// scalac: -Werror
+object Test {
+  import pack._
+
+  // java.util.List not java.awt.List or pack.List
+  Foo.list().listIterator()
+}

--- a/test/files/pos/t7232f/C.java
+++ b/test/files/pos/t7232f/C.java
@@ -1,0 +1,8 @@
+
+import p.D;
+
+import java.util.function.Supplier;
+
+public class C {
+    public final D d = new D();
+}

--- a/test/files/pos/t7232f/D.java
+++ b/test/files/pos/t7232f/D.java
@@ -1,0 +1,4 @@
+
+public class D {
+    public void notok() { }
+}

--- a/test/files/pos/t7232f/Test.scala
+++ b/test/files/pos/t7232f/Test.scala
@@ -1,0 +1,8 @@
+
+object Test {
+  def main(args: Array[String]): Unit = {
+    val c = new C()
+    val x = c.d
+    x.ok()
+  }
+}

--- a/test/files/pos/t7232f/q.scala
+++ b/test/files/pos/t7232f/q.scala
@@ -1,0 +1,6 @@
+
+package p
+
+class D {
+  def ok() = ()
+}

--- a/test/files/pos/t7232g/Foo.java
+++ b/test/files/pos/t7232g/Foo.java
@@ -1,0 +1,8 @@
+package pack;
+
+import java.util.*;
+
+public class Foo {
+    // should be java.util.List obviously.
+    public static List list() { throw new Error(); }
+}

--- a/test/files/pos/t7232g/Test.scala
+++ b/test/files/pos/t7232g/Test.scala
@@ -1,0 +1,6 @@
+// scalac: -Werror
+object Test {
+  import pack._
+
+  Foo.list().listIterator()
+}

--- a/test/files/run/t12195/C.java
+++ b/test/files/run/t12195/C.java
@@ -1,0 +1,8 @@
+
+import p.D;
+
+import java.util.function.Supplier;
+
+public class C {
+    public final D d = new D();
+}

--- a/test/files/run/t12195/D.java
+++ b/test/files/run/t12195/D.java
@@ -1,0 +1,3 @@
+
+public class D {
+}

--- a/test/files/run/t12195/Test.scala
+++ b/test/files/run/t12195/Test.scala
@@ -1,0 +1,13 @@
+
+object Test {
+  def main(args: Array[String]): Unit = {
+    val c = new C()
+    val x = c.d
+    assert(x.toString == "ok")
+  }
+}
+/* Because scalac used wrong class D in C.java, was:
+ * java.lang.NoSuchFieldError: d
+ *   at Test$.main(Test.scala:5)
+ *   at Test.main(Test.scala)
+ */

--- a/test/files/run/t12195/q.scala
+++ b/test/files/run/t12195/q.scala
@@ -1,0 +1,6 @@
+
+package p
+
+class D {
+  override def toString = "ok"
+}

